Allow initializing MGLGeoJSONSource from a shape

frederoni opened the pull request:
First stab at #6986 

I'm slightly hesitant about replacing `-[MGLGeoJSONSource initWithIdentifier:features:options:]` with `feature` because of the additional initialization of MGLShapeCollectionFeature devs has to do now compared to a concise array literal prior to this change.

@1ec5 👀

1ec5 commented:
I share you hesitation about removing `-[MGLGeoJSONSource initWithIdentifier:features:options:]`. `-[MGLShapeCollectionFeature shapeCollectionWithShapes:]` is kind of unwieldy. Should we ease the pain by keeping `-[MGLGeoJSONSource initWithIdentifier:features:options:]` as yet another initializer on MGLGeoJSONSource?

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.mm:
```diff
@@ -79,21 +80,43 @@ - (void)commonInit
+        id serialized = [NSJSONSerialization JSONObjectWithData:self.geoJSONData options:0 error:nil];
+        BOOL isFeatureCollection = [serialized isKindOfClass:NSArray.class] || ([serialized isKindOfClass:NSDictionary.class] && [[(NSDictionary *)serialized objectForKey:@"type"] isEqualToString:@"FeatureCollection"]);
+        
+        if (isFeatureCollection) {
```
Why does this conditional exist? `mbgl::style::Source::setGeoJSON()` takes a `mapbox::geojson::geojson`, which is the type returned by `mapbox::geojson::parse()`, so no `get()` is needed for either case. `_feature` is of type `id <MGLFeature>`, so `MGLFeatureFromMBGLFeature()` works for both cases too.

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.mm:
```diff
@@ -79,21 +80,43 @@ - (void)commonInit
-        for (id <MGLFeaturePrivate> feature in self.features) {
-            featureCollection.push_back([feature mbglFeature]);
+        // Feature collection
+        if ([self.feature isMemberOfClass:MGLShapeCollectionFeature.class]) {
```
MGLShapeCollectionFeature implements `-mbglFeature`, so this case is also unnecessary.

frederoni commented:
Removed debug code and cleaned up.
Left todo:
 - [ ] Support initializing MGLGeoJSONSource with a geometry?
 - [ ] `-[MGLGeoJSONSourceTests testMGLGeoJSONSourceWithPolygonFeatures]` is failing unexpectedly.
 - [ ] features are not set if `MGLGeoJSONSource` is initialized with data.

frederoni commented on platform/darwin/src/MGLGeoJSONSource.mm:
```diff
@@ -82,9 +93,8 @@ - (void)commonInit
-        const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature_collection>();
+        const auto geojson = mapbox::geojson::parse(string.UTF8String);
         source->setGeoJSON(geojson);
-        _features = MGLFeaturesFromMBGLFeatures(geojson);
```
If we're gonna support plain geometries, we can't assign features here.
Wouldn't it make sense to rename features to geometries and write a GeoJSON evaluator which overloads  featureCollection, feature and geometry?

1ec5 replied:
We need a `shape` property that can be any MGLShape (or any MGLFeature). We can have a `features` property as a convenience for when `shape` happens to contain an MGLShapeCollectionFeature.

frederoni commented:
Depending on what we want, I think this is good for another review and possibly also merge.
#6986 doesn't occur anymore but there's a drawback. Using `-[MGLGeoJSONSource initWithIdentifier:geoJSONData:options:]` will leave the features property unset because it can contain plain geometries.

1ec5 commented:
Maybe that means it needs to be a `shape` property (normally holding a shape collection feature), with `features` being a convenience property based on it?

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.h:
```diff
@@ -99,6 +99,19 @@ extern const MGLGeoJSONSourceOption MGLGeoJSONSourceOptionSimplificationToleranc
+ @param options An `NSDictionary` of options for this source.
+ @return An initialized GeoJSON source.
+ */
+- (instancetype)initWithIdentifier:(NSString *)identifier feature:(nullable id<MGLFeature>)feature options:(nullable NS_DICTIONARY_OF(MGLGeoJSONSourceOption, id) *)options NS_DESIGNATED_INITIALIZER;
```
Since `mbgl::style::GeoJSONSource::setGeoJSON()` takes an `mbgl::GeoJSON` (a variant of `geometry`, `feature`, or `feature_collection`), it would be better for this initializer to take any MGLShape. All MGLShape subclasses in turn have subclasses that conform to MGLFeature, so we wouldn’t be losing any expressiveness that way.

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.h:
```diff
@@ -114,7 +128,7 @@ extern const MGLGeoJSONSourceOption MGLGeoJSONSourceOptionSimplificationToleranc
  `-[MGLStyle addSource]`.
  */
-@property (nonatomic, nullable) NS_ARRAY_OF(id <MGLFeature>) *features;
+@property (nonatomic, nullable) MGLShape *shape;
```
Update the documentation for this property.

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.h:
```diff
@@ -138,6 +152,10 @@ extern const MGLGeoJSONSourceOption MGLGeoJSONSourceOptionSimplificationToleranc
 @property (nonatomic, nullable) NSURL *URL;
 
+/**
+ Returns all objects from the shape that conforms to `MGLFeature`.
```
This makes it sound like it does a deep search through the shape collection feature for any shapes, whereas it's only set to the top-level features, and only if a shape collection feature or array of features was passed in on initialization.

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.h:
```diff
@@ -138,6 +152,10 @@ extern const MGLGeoJSONSourceOption MGLGeoJSONSourceOptionSimplificationToleranc
+/**
+ Returns all objects from the shape that conforms to `MGLFeature`.
+ */
+- (NSArray<MGLFeature> *)features;
```
Use `NS_ARRAY_OF()` in public APIs for the benefit of developers compiling against the iOS 8 or OS X 10.10 SDKs.

Also, this method should be a property like before.

Finally, the return type should be an array of `id <MGLFeature>`, not an NSArray conforming to the MGLFeature protocol. There's a big difference.

1ec5 commented on platform/darwin/src/MGLPolygon.mm:
```diff
@@ -38,23 +38,19 @@ - (instancetype)initWithCoordinates:(CLLocationCoordinate2D *)coords count:(NSUI
-    }
 
-    mbgl::FillAnnotation annotation { geometry };
+    auto polygon = mbgl::Polygon<double> { [self ring] };
```
What about interior rings?

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.mm:
```diff
@@ -155,41 +171,47 @@ - (void)setGeoJSONData:(NSData *)geoJSONData
+        MGLShapeCollection *collection = (MGLShapeCollection *)_shape;
+        for (MGLShape *shape in collection.shapes) {
+            if ([shape conformsToProtocol:@protocol(MGLFeature)]) {
+                [features addObject:(id <MGLFeature>)shape];
```
Would it be more efficient to compute the `features` property as soon as `shape` is set, or on demand when calling `-features`? Or perhaps a lazy-loading strategy is called for, similar to how I treated `overlayBounds` in #7251.

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.mm:
```diff
@@ -49,9 +50,20 @@ - (instancetype)initWithIdentifier:(NSString *)identifier URL:(NSURL *)url optio
     return self;
 }
 
+- (instancetype)initWithIdentifier:(NSString *)identifier feature:(id<MGLFeature>)feature options:(NSDictionary<MGLGeoJSONSourceOption,id> *)options
```
Any reason to require a feature instead of any ordinary shape? Looks like we’re only using this parameter for an MGLShape-typed property anyways.

frederoni replied:
Mostly not to break the API. Should we rename it to `-[MGLGeoJSONSource initWithIdentifier:shape:options:]` or keep `feature:` and `features:` while adding `shape:` ?

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.h:
```diff
@@ -93,28 +94,29 @@ extern const MGLGeoJSONSourceOption MGLGeoJSONSourceOptionSimplificationToleranc
  @return An initialized GeoJSON source.
  */
-- (instancetype)initWithIdentifier:(NSString *)identifier features:(NSArray<id<MGLFeature>> *)features options:(nullable NS_DICTIONARY_OF(MGLGeoJSONSourceOption, id) *)options NS_DESIGNATED_INITIALIZER;
+- (instancetype)initWithIdentifier:(NSString *)identifier shape:(nullable MGLShape *)shape options:(nullable NS_DICTIONARY_OF(MGLGeoJSONSourceOption, id) *)options NS_DESIGNATED_INITIALIZER;
```
What does it mean to create a source from no shapes at all? It does work in the current release branch, using an empty array of features, but we should make sure it works when shape is nil.

frederoni replied:
nil shape works but I can't think of any case where you'd want to do that.

1ec5 commented on platform/darwin/src/MGLGeoJSONSource.mm:
```diff
@@ -174,22 +168,20 @@ - (void)setURL:(NSURL *)URL
     self.rawSource->setGeoJSON(geojson);
     
-    _features = MGLFeaturesFromMBGLFeatures(featureCollection);
+    [self willChangeValueForKey:@"shape"];
```
These change notifications are unnecessary. Since setShape: is a standard KVC selector, these change notifications will be generated for you implicitly.

1ec5 commented:
The Coveralls failure is caused by an unrelated change in test coverage in mbgl. You can get around it by pushing another change, so that coverage will be even.

frederoni commented on platform/darwin/test/MGLGeoJSONSourceTests.mm:
```diff
@@ -236,7 +242,8 @@ - (void)testMGLGeoJSONSourceWithShapeCollectionFeatures {
 
-    XCTAssertThrowsSpecificNamed([[MGLGeoJSONSource alloc] initWithIdentifier:@"source-id" features:@[shapeCollectionFeature_1] options:nil], NSException, @"Method unavailable");
+#warning DEBUG
+    //XCTAssertThrowsSpecificNamed([[MGLGeoJSONSource alloc] initWithIdentifier:@"source-id" shape:shapeCollectionFeature_1 options:nil], NSException, @"Method unavailable");
```
@1ec5 what was the purpose of this test?

boundsj commented:
The iOS release branch now has https://github.com/mapbox/mapbox-gl-native/commit/c8e34d581b80beea247229051795a8ab3920c378 so if this branch is rebased on top of that, the node tests will pass.
